### PR TITLE
httpbakery/agent: export Agent type

### PR DESCRIPTION
**Pull request by rogpeppe**
We export the Agent type in preparation for providing
serialization support for Visitor. We also
make Agent the internal representation too
and use a slightly more efficient, allocation-free, lookup
algorithm.

### CODE REVIEW

File: httpbakery/agent/agent.go [hunk 1]
```diff
@@ -123,11 +124,21 @@ func LoginCookie(req *http.Request) (username string, key *bakery.PublicKey, err
...
+}
+
+// agent represents an agent that can be used for agent authentication.
+type Agent struct {
+	// URL holds the URL associated with the agent.
+	URL string
+	// Username holds the username to use for the agent.
+	Username string
+	// Key holds the agent's private key pair.
+	Key *bakery.KeyPair
```
Comment (mhilton) [inline httpbakery/agent/agent.go:141]: Using a KeyPair representation will be awkward if you want to use the same type for serializing as an agent-cookie.

Comment (mhilton) [reply httpbakery/agent/agent.go:141]: Also with existing serializations already in production (well staging)

**Review by mhilton — APPROVED** [review]
:+1:

File: httpbakery/agent/agent.go (hunk 1 above)
Comment (rogpeppe) [reply httpbakery/agent/agent.go:141]: I don't think we need to use the same type for the agent cookie. I see them as two different things, even though one can be derived from the other.

File: httpbakery/agent/agent.go [hunk 2]
```diff
@@ -136,98 +147,151 @@ type Visitor struct {
...
+
+// AddAgent adds an agent to the visitor. The agent information will be
+// used when sending discharge requests to all URLs under the given URL.
+// If more than one agent matches a target URL then the agent with the
+// most specific matching URL will be used. Longer paths are counted as
+// more specific than shorter paths.
+//
+// Unlike HTTP cookies, a trailing slash is not significant, so for
+// example, if an agent is registered with the URL
+// http://kremvax.com/foo, its information will be sent to
```
Comment (jrwren) [inline httpbakery/agent/agent.go]: Maybe use example.com here instead of kremvax.com.

**Review by jrwren — APPROVED** [review]

File: httpbakery/agent/agent.go (hunk 2 above)
Comment (rogpeppe) [reply httpbakery/agent/agent.go]: done

**Pull request merged**